### ISSUE TITLE
parallel_for failed: cudaErrorInvalidDevice: invalid device ordinal

### ISSUE THREAD

**Issue by ca1773130n** [MrNeRF/LichtFeld-Studio#792]
@maintainers, thanks for the great work and active development.

I'm trying to use LFS in headless mode on a server, but I'm getting the error (log below).

OS: Ubuntu 24.04
GPU: single H200 (140GB)

[16:02:50.627] [info] argument_parser.cpp:488  Scaling training steps by factor: 1
[16:02:50.627] [info] main.cpp:30  LichtFeld Studio
[16:02:50.627] [info] main.cpp:31  version v0.3.0-195-g99ce575e | tag 99ce575e
[16:02:50.627] [info] application.cpp:75  CUDA driver version: 12.8
[16:02:50.627] [info] application.cpp:37  Starting headless training...
[16:02:50.627] [info] training_setup.cpp:65  Loading dataset from: {BLENDER_NERF_DATASET_PATH}
[16:02:50.631] [info] loader_service.cpp:77  Using Blender/NeRF loader for: {BLENDER_NERF_DATASET_PATH}
[16:02:50.632] [info] blender_loader.cpp:154  Loading Blender/NeRF dataset from: {BLENDER_NERF_DATASET_PATH}/transforms.json
[16:02:51.466] [info] transforms.cpp:351  Loaded 1161 cameras from transforms file
[16:02:51.516] [info] gpu_slab_allocator.hpp:168  GPUSlabAllocator: 288.00 MB allocated
[16:03:00.590] [info] dataset.hpp:210  Dataset created with 1161 images (split: 2)
[16:03:06.813] [info] transforms.cpp:516  Successfully loaded PLY point cloud with 11640777 points
[16:03:06.843] [info] blender_loader.cpp:250  Loaded 11640777 points from dense_pc.ply
[16:03:06.843] [info] blender_loader.cpp:282  Blender/NeRF dataset loaded successfully in 16211ms
[16:03:06.843] [info] blender_loader.cpp:283    - 1161 cameras
[16:03:06.851] [info] training_setup.cpp:71  Dataset loaded successfully using Blender/NeRF loader
[16:03:06.851] [info] training_setup.cpp:178  Adding point cloud to scene (11640777 points) - SplatData will be created when training starts
[16:03:06.956] [info] training_setup.cpp:244  Loaded dataset 'nerfstudio' into scene: 1161 train cameras (1161 with masks)
[16:03:19.640] [info] splat_data.cpp:870  Scene scale: 73.61429
[16:03:19.640] [info] splat_data.cpp:871  Initialized SplatData: 11640777 points, max SH degree: 3, SH coefficients: 16, sh0 shape: [11640777, 1, 3], shN shape: [11640777, 15, 3]
[16:03:19.640] [info] training_setup.cpp:409  Created training model with 11640777 gaussians
[16:03:19.886] [info] trainer.cpp:345  Initializing trainer with 4500 iterations
[16:03:19.887] [info] dataset.hpp:210  Dataset created with 1161 images (split: 2)
[16:03:19.887] [info] trainer.cpp:391  Using all 1161 images for training (no evaluation)
[16:03:19.887] [info] mcmc.cpp:574  Pre-allocating capacity for 20000000 Gaussians (current size: 11640777, utilization: 58.2%)
[16:03:19.906] [info] mcmc.cpp:603  Pre-allocated capacity: 11640777/20000000 Gaussians (58.2%)
[16:03:19.907] [info] strategy_utils.cpp:59  AdamOptimizer: pre-allocating capacity for 20000000 Gaussians (optimizer states)
[16:03:19.911] [info] mcmc.cpp:628  MCMC strategy initialized with 11640777 Gaussians
[16:03:19.915] [info] trainer.cpp:472  Background color set to RGB(0.00, 0.00, 0.00)
[16:03:19.915] [info] trainer.cpp:553  Visualization: disabled
[16:03:19.915] [info] trainer.cpp:554  Strategy: mcmc
[16:03:19.915] [info] trainer.cpp:564  Trainer initialization complete
[16:03:19.918] [info] trainer.cpp:1338  Starting training loop
[16:03:19.919] [info] pipelined_image_loader.cpp:99  [PipelinedImageLoader] batch_size=8, prefetch=8, io_threads=2, cold_threads=1
[16:03:19.920] [info] nvcodec_image_loader.cpp:213  [nvImageCodec Diagnostics] === Starting nvImageCodec availability check ===
[16:03:19.920] [info] nvcodec_image_loader.cpp:90  [nvImageCodec Diagnostics] CUDA device count: 1
[16:03:19.921] [info] nvcodec_image_loader.cpp:95  [nvImageCodec Diagnostics] GPU 0: NVIDIA H200 (SM 9.0, 143166 MB, compute capability 9.0)
[16:03:19.921] [info] nvcodec_image_loader.cpp:108  [nvImageCodec Diagnostics] GPU 0 compute capability 9.0 - nvJPEG hardware decode should be available
[16:03:19.921] [info] nvcodec_image_loader.cpp:117  [nvImageCodec Diagnostics] CUDA driver version: 12.8
[16:03:19.921] [info] nvcodec_image_loader.cpp:124  [nvImageCodec Diagnostics] CUDA runtime version: 12.8
[16:03:19.922] [info] nvcodec_image_loader.cpp:136  [nvImageCodec Diagnostics] Extensions directory: /data/workspace_294/edward/project/spatial_roadview/experiments/trainers/LichtFeld-Studio/build/extensions
[16:03:19.923] [info] nvcodec_image_loader.cpp:175  [nvImageCodec Diagnostics] Found 1 extension files:
[16:03:19.923] [info] nvcodec_image_loader.cpp:177  [nvImageCodec Diagnostics]   - libnvjpeg_ext.so
[16:03:19.923] [info] nvcodec_image_loader.cpp:232  [nvImageCodec Diagnostics] Attempting to create nvImageCodec instance...
[16:03:19.927] [info] nvcodec_image_loader.cpp:259  [nvImageCodec Diagnostics] Instance creation with BUILTIN modules: SUCCESS
[16:03:19.954] [info] nvcodec_image_loader.cpp:274  [nvImageCodec Diagnostics] Instance creation with EXTENSION modules: SUCCESS
[16:03:19.985] [info] nvcodec_image_loader.cpp:309  [nvImageCodec Diagnostics] Decoder creation: SUCCESS
[16:03:19.986] [info] nvcodec_image_loader.cpp:317  [nvImageCodec Diagnostics] === nvImageCodec AVAILABLE ===
[16:03:19.987] [info] pipelined_image_loader.cpp:128  [PipelinedImageLoader] Started 2 I/O, 1 GPU, 1 cold threads
[16:03:19.993] [info] nvcodec_image_loader.cpp:402  [NvCodecImageLoader] Initializing: device=0, pool=8, fallback=true
[16:03:20.090] [info] nvcodec_image_loader.cpp:447  [NvCodecImageLoader] 8 decoders ready
[16:03:20.255] [info] memory_arena.cu:430  VMM initialized: device=0, virtual=32 GB
[16:03:20.358] [info] pipelined_image_loader.cpp:140  [PipelinedImageLoader] Shutting down...
[16:03:20.390] [info] pipelined_image_loader.cpp:167  [PipelinedImageLoader] Done: 3 loaded, 0 hits, 8 misses
[16:03:20.391] [error] application.cpp:60  Training error: Training step failed: parallel_for failed: cudaErrorInvalidDevice: invalid device ordinal
[16:03:20.391] [info] image_io.cpp:544  [BatchImageSaver] Starting with 4 worker threads
Training [                                        ] 100% [00m:00s<00m:00s] Initializing...                                                                                                                          
Training [                                        ] 100% [00m:00s<00m:00s] Initializing...                                                                                                                          

[16:03:20.400] [warn] memory_arena.cu:360  Emergency cleanup started
[16:03:20.401] [warn] memory_arena.cu:388  Emergency cleanup completed
[16:03:20.437] [info] image_io.cpp:558  [BatchImageSaver] Shutting down...
[16:03:20.437] [info] image_io.cpp:570  [BatchImageSaver] Shutdown complete
[16:03:20.437] [info] memory_arena.cu:982  Arena stats: committed=1002 MB, peak=982 MB (98.0%), frames=1, reallocs=2, runtime=0s

**Comment by MrNeRF** [comment]
I am seeing that your initial point cloud  11,640,777 is quite large. I haven't tested myself with such a huge initial pc. 
It would be awesome if I could get my hands on this dataset to debug the issue.

**Comment by ca1773130n** [comment]
@MrNeRF 
Thanks. I tried not giving initial points but it's the same.

...same logs...
[17:57:03.543] [warn] blender_loader.cpp:254  No PLY found, using 10000 random points
...same logs...
[17:57:04.257] [error] application.cpp:60  Training error: Training step failed: parallel_for failed: cudaErrorInvalidDevice: invalid device ordinal

I'm investigating this issue now. I will post here once I fix it.

**Comment by claarbelgium** [comment]
We were just starting to test LichtFeld Studio on our workstations here but we also get this same error.
Our dataset is made with 3DS max, we have 900 images, and created a transforms.json to load the camera's. The position & directions all seem to be correct in the viewport, and it also says it's using 10000 random points. When we want to start the training this same error comes up: "parallel_for failed: cudaErrorInvalidDevice: invalid device ordinal"
We have tried on two different workstations one with RTX 2070 SUPER and a RTX 4080 and we have tried with multiple versions of the nvidia drivers but nothing that fixed this yet.
We do see a cloudy cube appearing at the center of the original xy axis (maybe the 10000 random points?), but our camera's are further from the origin. Maybe that has something to do with it? 

Update; We scaled and moved the camera's so they fit within the box and it's processing now!

**Comment by ca1773130n** [comment]
@claarbelgium 
I don't think the root cause of this problem is related to datasets.
I haven't figure it out yet but parallel_for() is used only in these codes:
- src/io/formats/ply.cpp 
- build/vcpkg_installed/x64-linux/include/oneapi/tbb
- build/vcpkg_installed/x64-linux/include/OpenImageIO/parallel.h

I've tested by adding codes to ply.cpp for forcing cuda device 0 but the result was the same, meaning the file is ok.
So it seems the problem is due to TBB or OpenImageIO.

**Comment by ca1773130n** [comment]
@claarbelgium 
Thanks for the update. Does it mean that we could just ignore the parallel_for error when training?

**Comment by claarbelgium** [comment]
@ca1773130n Yes we assume so. The error disappeared completely as soon as we positioned the camera's exactly at the center and in a custom scale, the random 10000 points are covering now approximately the same volume to train on and it's working.

**Comment by MrNeRF** [comment]
Can you please try either the latest commit or the latest binary build? I think this could be solved.

**Comment by ca1773130n** [comment]
@MrNeRF 
I tested the latest build and it works without the parallel_for() error with my dataset.

[22:52:24.851] [info] nvcodec_image_loader.cpp:309  [nvImageCodec Diagnostics] Decoder creation: SUCCESS
[22:52:24.853] [info] nvcodec_image_loader.cpp:317  [nvImageCodec Diagnostics] === nvImageCodec AVAILABLE ===
[22:52:24.853] [info] pipelined_image_loader.cpp:128  [PipelinedImageLoader] Started 2 I/O, 1 GPU, 2 cold threads
[22:52:24.855] [info] nvcodec_image_loader.cpp:402  [NvCodecImageLoader] Initializing: device=0, pool=8, fallback=true
[22:52:24.948] [info] nvcodec_image_loader.cpp:447  [NvCodecImageLoader] 8 decoders ready
[22:52:25.166] [info] memory_arena.cu:430  VMM initialized: device=0, virtual=32 GB
Training [████████████████████████▌░░░░░░░░░░░░░░░] 60% [00m:41s<00m:27s] 2600/4500 | Loss: 0.1778 | Splats: 17198724 (+)   

Closing this issue. Thanks for the update! congrats LFS team

**Comment by ben-williams-ai** [comment]
For others encountering this issue who don't want to run a full upgrade yet: I have found increasing the `init_scaling` parameter in the config helps. By defeault it is 0.1, I have found on three very different datasets that increasing it somewhere between 0.1 - 0.5 always seems to fix this. I am not sure why and whether adjusting other paramters could also provide a fix. I believe this parameter controls how large the initial gaussians are. I am always passing in sparse point clouds so not sure if that is related.
